File: Utilities/linked_list.py
# Node structure for linked list
class Node:
    def __init__(self, val):
        self.data = val
        self.next = None

# Linked list implementation 
class Linked_list:
    def __init__(self):
        self.head = None
    
    # Insert at end
    def insert(self, val):
        if self.head == None:
            self.head = Node(val)
        else:
            ptr = self.head
            while ptr.next != None:
                ptr = ptr.next
            ptr.next = Node(val)
        
class Stack:
    def __init__(self):
        self.head = None
    
    # Insert at beginning
    def push(self, new_node:Node):
        if self.head:
            new_node.next = self.head
        self.head = new_node
        return self.head
    
    # Delete at beginning
    def pop(self):
        if self.head:
            popped = self.head
            self.head = self.head.next
            popped.next = None
            return popped.data
        else:
            print("Stack Underflow")
            return None

class Queue:
    def __init__(self):
        self.head = None
        self.tail = None
    
    # Insert at end(tail)
    def push(self, new_node:Node):
        if self.head == None:
            self.head = self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        return self.tail
    
    # Delete at beginning
    def pop(self):
        if self.head:
            popped = self.head
            self.head = self.head.next
            popped.next = None
            return popped.data
        else:
            print("Queue Underflow")
            return None

File: Utilities/linked_list.cpp
#include <bits/stdc++.h>

using namespace std;

// Contains data for a single node in linked list
class Node {
    public: 
    int data;
    Node *next;
    Node() {
        data = 0;
        next = NULL;
    }
    Node(int n) {
        data = n;
        next = NULL;
    }
};

// Linked list implementation class
class SinglyLinkedList {
    public:
    Node *head;
    int size;

    SinglyLinkedList() {
        size = 0;
        head = NULL;
    }
    SinglyLinkedList(int data) {
        size = 1;
        head = new Node(data);
    }
    SinglyLinkedList(Node *node) {
        size = 1;
        head = node;
    }
    
    // Insert a data at end
    void PushBack(int data) {
        Node *newNode = new Node(data);
        if(head == NULL) {
            head = newNode;
        }
        else {
            Node *temp = head;
            while(temp->next != NULL) {
                temp = temp->next;
            }
            temp->next = newNode;
        }
        size++;
    }

    // Add a new node to the front of the list
    void PushFront(int data) {
        Node *newNode = new Node(data);
        newNode->next = head;
        head = newNode;
        size++;
    }

    // Pop the last node
    int PopLast() {
        Node *temp = head;
        Node *delNode;
        int deleted;
        while(temp->next->next != NULL) {
            temp = temp->next;
        }
        delNode = temp->next;
        temp->next = NULL;
        deleted = delNode->data;
        // free memory
        delete delNode;
        size--;
        return deleted;
    }

    // Pop node at index - index is 0 based
    int PopAt(int index) {
        if(index >= size) {
            return;
        }
        Node *temp = head;
        Node *delNode;
        int deleted;
        for(int i=0; i<index; i++) {
            temp = temp->next;
        }
        delNode = temp->next;
        temp->next = temp->next->next;
        deleted = delNode->data;
        delete delNode;
        size--;
        return deleted;
    }

    // Print the linked list
    void PrintList() {
        Node *temp = head;
        while(temp != NULL) {
            cout << temp->data << ' ';
            temp = temp->next;
        }
        cout << endl;
    }
};

int main() {
    int n, data;
    cin >> n;
    SinglyLinkedList list;
    for(int i=0; i<n; i++) {
        cin >> data;
        list.PushBack(data);
    }
    list.PrintList();
    list.PushFront(-3);
    cout << list.PopLast() << endl;
    list.PrintList();
}